routerrpc: TrackPayments

JssDWt opened the pull request:
## Change Description
An RPC method to subscribe to _any_ paymentattempt. There is currently already a method for subscribing to updates for a single payment (TrackPaymentV2). This PR adds a method to subscribe to updates for any payment.

Note: This is my first contribution to LND.

## Motivation
There's a lot of relevant information about the channel graph in payment events that can be used to construct heuristics about nodes and channels in the network. This is important information for a routing node. Especially in the potential future, if trampoline nodes become a thing. There's many tools out there today that can help send a payment, like RTL, Thunderhub, other dashboards, LNDg, balanceofsatoshis, connected wallets like Zeus/BlueWallet and self-coded stuff. If multiple tools are used to send payments, it's hard to monitor the paymentattempts currently. You'd have to use a combination of `ListPayments` and `TrackPaymentV2` to collect the relevant information and depending on your polling interval on `ListPayments`, you might miss a few. Especially now that balanceofsatoshis/lnservice deletes payments on the fly. This method makes it a lot easier to build a monitoring tool that collects information from payment attempts made. I've been struggling with this myself and I've seen some other cases in the wild where people were missing a feature to subscribe to payments.

## Steps to Test
- Subscribe to payments 
   `curl -X GET --cacert "$LND_DIR/tls.cert" --header "$MACAROON_HEADER" https://localhost:8080/v2/router/payments`
- Make a payment from your node

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

joostjager commented on routing/control_tower.go:
```diff
@@ -255,6 +261,27 @@ func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
+		subscriber.queue.ChanIn() <- payment
+	}
+
+	p.subscribersMtx.Lock()
```
You may want to think about race conditions here. If a new payment becomes pending right at this point, it will probably be missed on the stream?

JssDWt replied:
Yeah, I thought about race conditions here. It's basically a tradeoff between three things:
1) Ignore inflight payments that started before the subscription
2) Do include the inflight payments, but maybe miss an update in between around the start of the subscription (the way it's coded now)
3) Do include the inflight payments and all updates in between, but potentially out of order at the start of the subscription and with potential duplicates. (if we'd add the subscription to the list before fetching inflight payments)

I preferred number 2 myself, that's why it's coded like this.

joostjager replied:
I think that 3 is the ideal option, closest to what the user would expect. Maybe it is worth thinking about this a bit more, especially because there are no missed updates for the single payment subscription (facilitated by `paymentsMtx`).

Maybe a similar solution with a lock is possible for `SubscribeAllPayments`? The updates go into a memory queue, so it won't block on rpc clients not being responsive.

JssDWt replied:
I've made an attempt to fix that. Judging from the comments in the code, the payments mutex is there to to do the exact same thing I'm trying to do. But since this change cannot lock on a specific payment hash, we'd need some other locking mechanism.

_If_ the payments mutex is there to avoid payment updates from happening _while_ we're setting up a new subscription, we can achieve a pretty efficient mechanism with a reader-writer lock, as I did in my latest commit. 

A problem could be that the underlying code to update the payments in the database is actually relying on the mutex on the specific payment hash, so no 2 database updates can actually happen at the same time. That's a side-effect of the existing mutex, but something could rely on it. 

In this new scheme, no database updates can happen _while_ we're making a new subscription, but during normal flow, database updates could occur in parallel.

What do you think?

Another option would be 2 mutexes. One for the payment hash (to avoid concurrent database updates) and a reader-writer mutex while taking the subscription (to avoid updates while making the subscription). But it could be hard to follow what's going on as a reader of the code then.

joostjager replied:
I am not sure if it is safe. With only reader locks, I think the order in which db updates are applied isn't guaranteed to be reflected in the subscriber events for a single payment. Maybe if every payment is only updated from a single thread, it is not a problem.

An additional mutex that just prevents events from being missed may be the safer option.

Nice analysis btw.

joostjager replied:
Phew, lots of locks. I went through the code, but have to say that it isn't easy to see how all these combinations of locks, rlocks and unlocks work together and that it is 100% correct. Part of that is pre-existing, it wasn't simple to start with.

Not sure what to do about it though. It may be good to take another look at what exactly the critical path is that we need to optimize for. The expensive operations that prevent us from just wrapping everything in a single lock that kills all concurrency. And from there, see if there are other options.

JssDWt replied:
Yeah it makes it very difficult to read indeed.
It could be better to 
- drop the additional locking
- Add the subscriber before fetching the inflight payments

That way the user might get some duplicate events and potentially out of order events at the beginning of the stream, but it saves a lot of complexity in the code here.

Another option is I can merge that subscribersMtx and newSubscriptionMtx maybe. I'll have a look at the exact tradeoffs there.

joostjager replied:
Yes, maybe it is not worth the extra complexity. This might be a good time for another reviewer to take a look at it. It seems the options are quite well laid out now.

joostjager commented on routing/control_tower.go:
```diff
@@ -57,6 +57,10 @@ type ControlTower interface {
 	// sent out immediately.
 	SubscribePayment(paymentHash lntypes.Hash) (*ControlTowerSubscriber,
 		error)
+
```
General comment about commit message: typically they are prefixed with the subsystem(s).

JssDWt replied:
Fixed

joostjager commented on routing/control_tower.go:
```diff
@@ -306,20 +306,40 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
+	}
+}
+
+func notifySubscriber(subscriber *ControlTowerSubscriber,
```
If you do this extraction in a separate commit, the diff where 'all' is introduced becomes easier to read.

JssDWt replied:
I removed this extraction, it's a lot easier to read without the extraction indeed.

joostjager commented on lnrpc/routerrpc/router.proto:
```diff
@@ -22,6 +22,12 @@ service Router {
+    /*
+    SubscribePayments returns an update stream for every payment.
+    */
+    rpc SubscribePayments (SubscribePaymentsRequest)
```
For consistency (within this sub-server...), should it be `TrackPayments` or `TrackAllPayments`?

JssDWt replied:
Yes, probably right, updated the name

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -771,11 +771,66 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+	// Stream updates back to the client.
+	for {
+		select {
+		case item, ok := <-subscription.Updates:
```
Is there potential for code reuse here from the single payment subscription?

JssDWt replied:
It's the interfaces `Router_TrackPaymentV2Server` and `Router_TrackPaymentsServer`
So `stream.Send(rpcPayment)` looks the same, but it's actually two different things.
That got me to copy the thing.

Another option would be to merge TrackPaymentV2 and TrackPayments completely, by allowing the passed hash to be null. But I don't think that's a good idea because they do different things. TrackPayments tracks any currently in-flight payment, while TrackPaymentV2 gets you the status of a specific payment, regardlesss whether it's in-flight or not.

joostjager replied:
No, I wouldn't overload the existing rpc method. But you can extract common code by passing references to the `Send` function and `Done` channel perhaps?

JssDWt replied:
Done some code reuse there

joostjager commented on routing/control_tower.go:
```diff
@@ -58,8 +58,9 @@ type ControlTower interface {
-	// the current state of every inflight payment is always sent out immediately.
+	// SubscribeAllPayments subscribes to updates for all payments. A first
+	// update with the current state of every inflight payment is always sent
+	// out immediately.
```
You want to squash this with the commit that introduced it

joostjager commented on docs/release-notes/release-notes-0.15.0.md:
```diff
@@ -92,6 +92,10 @@
   change, it allows encrypted failure messages to be returned to the sender.
   Additionally it is possible to signal a malformed htlc.
 
+* Add [SubscribePayments](https://github.com/lightningnetwork/lnd/pull/6335)
```
Maybe nice to mention in the PR description some use cases for this new functionality.

JssDWt replied:
Done!

joostjager replied:
Nice addition!

JssDWt commented:
Thanks for your review @joostjager. All valid comments, I'll work on modifying those points and improve the commit structure.

JssDWt commented on routing/control_tower.go:
```diff
@@ -291,9 +291,9 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
-
-	if len(list) == 0 {
+	subscribersAllPayments := p.subscribersAllPayments
+	subscribersPaymentHash, ok := p.subscribers[paymentHash]
```
Fixed

JssDWt commented on routing/control_tower.go:
```diff
@@ -306,20 +306,40 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
-
-		// If subscriber disappeared, skip notification. For further
-		// notifications, we'll keep skipping over this subscriber.
 		case <-subscriber.quit:
```
Fixed by removing the extraction

JssDWt commented on lnrpc/routerrpc/router.proto:
```diff
@@ -22,6 +22,12 @@ service Router {
     rpc TrackPaymentV2 (TrackPaymentRequest) returns (stream lnrpc.Payment);
 
+    /*
+    SubscribePayments returns an update stream for every payment.
```
Fixed

joostjager commented on routing/control_tower.go:
```diff
@@ -57,6 +57,11 @@ type ControlTower interface {
 	// sent out immediately.
 	SubscribePayment(paymentHash lntypes.Hash) (*ControlTowerSubscriber,
 		error)
+
```
Nit about commit message: generally the sentence after the subsystem indicator isn't capitalized.

JssDWt replied:
Right, didn;t notice that, updated!

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -771,11 +775,63 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
 
+		case <-s.quit:
+			return errServerShuttingDown
+		case <-stream.Context().Done():
```
newline above case

JssDWt replied:
Fixed

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -740,10 +744,57 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+		noInflightUpdates,
+		stream.Send,
+		stream.Context,
+		fmt.Sprintf("Payment stream %v canceled", identifier),
```
I think this log can also be done outside the call by checking the return value of `trackPaymentStream` for `context.Done`

joostjager commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -740,10 +744,57 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+}
+
+// trackPaymentStream streams payment updates to the client.
+func trackPaymentStream(s *Server,
```
Maybe if you make this a method of `Server`, you don't need to pass `s` and `router`.

JssDWt replied:
Good, done

JssDWt commented on routing/control_tower.go:
```diff
@@ -131,6 +143,8 @@ func (p *controlTower) RegisterAttempt(paymentHash lntypes.Hash,
 	attempt *channeldb.HTLCAttemptInfo) error {
 
 	p.paymentsMtx.Lock(paymentHash)
+	p.newSubscriptionMtx.RLock()
```
For the new subscription to `all` payments you only want to make sure that there are no updates while creating the new subscription, so you don't miss any. Ideally you would just lock the paymentsMtx when creating the 'all' subscription, but because that is hash-based, you can't lock all payments with that. That's why you need an additional mutex that can lock all payments at once. 

The payment mutex makes sure the updates for each individual payment occur in the order they are submitted to the database. So you're taking advantage of that locking in the 'all' subscription too.

JssDWt replied:
You don't want to `Lock` on a single mutex, because that would eliminate concurrent payments.

So the resulting stream to the user will have updates per payment that will be transmitted in the order they are stored in the database. But for different payments the order may vary. So the state in the database could be like this:
1) attempt on payment 1
2) attempt on payment 2

And in the update stream the user could see this:
1) attempt on payment 2
2) attempt on payment 1

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

JssDWt commented:
I think this PR is ready to be looked at by maintainers of LND. A decision maybe you guys can make is, should I use locks to make sure TrackPayments always returns exactly one update for every internal update, or is it better to not use the additional locks (difficult!) and allow duplicate updates if a payment is inflight while starting the subscription to Trackpayments. The alternative is posed here: https://github.com/lightningnetwork/lnd/pull/6335/files#r847158161

JssDWt commented:
I'm not sure how to request a review from someone other than Joost

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@jssdwt, remember to re-request review from reviewers when ready

JssDWt commented:
@guggero Is there anything I can do to make this PR easier to review?

guggero commented on routing/control_tower.go:
```diff
@@ -109,9 +115,10 @@ type controlTower struct {
-		subscribers: make(map[lntypes.Hash][]*ControlTowerSubscriber),
-		paymentsMtx: multimutex.NewHashMutex(),
+		db:                     db,
+		subscribersAllPayments: make([]*ControlTowerSubscriber, 0),
```
Slices generally don't need to be initialized, unless they should be pre-sized to a specific length to avoid re-sizing at runtime. So I think this line can just be removed.

JssDWt replied:
Done

guggero commented on routing/control_tower.go:
```diff
@@ -264,8 +293,9 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
 	p.subscribersMtx.Lock()
-	list, ok := p.subscribers[paymentHash]
-	if !ok {
+	subscribersAllPayments := p.subscribersAllPayments
```
I don't think there's a need to add a new reference to the struct variable to save 2 characters... Just makes things harder to read. So should use `p.subscribersAllPayments` everywhere in this method.

JssDWt replied:
Done

guggero commented on routing/control_tower.go:
```diff
@@ -294,4 +324,19 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
+			p.subscribersMtx.Lock()
+			p.subscribersAllPayments = append(
+				p.subscribersAllPayments[:index],
+				p.subscribersAllPayments[index+1:]...)
```
nit: put closing parenthesis on next line.

JssDWt replied:
Done

guggero commented on routing/control_tower_test.go:
```diff
@@ -182,6 +182,137 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
+	t.Parallel()
+
+	db, err := initDB()
+	if err != nil {
```
nit: new code should use the `require` library.

JssDWt replied:
Done, I hope I did this in a proper way

guggero commented on lnrpc/routerrpc/router.proto:
```diff
@@ -22,6 +22,11 @@ service Router {
     rpc TrackPaymentV2 (TrackPaymentRequest) returns (stream lnrpc.Payment);
 
+    /*
+    TrackPayments returns an update stream for every payment.
```
nit: every payment that is not in a terminal state?

JssDWt replied:
Updated

guggero commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -740,10 +744,55 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+
+	// Stream updates to the client
+	err = s.trackPaymentStream(
+		subscription,
```
style nit: we try to make method calls a bit more compact by fitting as many arguments on one line as possible (while respecting the max line length).

guggero replied:
Same below.

JssDWt replied:
Done

guggero commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -740,10 +744,55 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+		return err
+	}
+
+	// Stream updates to the client
```
nit: missing full stop.

JssDWt replied:
Done

guggero commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -760,23 +809,22 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
 			}
 
-			rpcPayment, err := router.MarshallPayment(result)
+			rpcPayment, err := s.cfg.RouterBackend.MarshallPayment(result)
```
style nit: line too long (when using 8 spaces for the tab character).

JssDWt replied:
Done, I indeed had 4 spaces as tabsize...

guggero commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -740,10 +744,55 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+	subscription *routing.ControlTowerSubscriber,
+	noInflightUpdates bool,
+	send func(*lnrpc.Payment) error,
+	context func() context.Context) error {
```
The context should remain the same throughout a call, shouldn't it? So instead of passing in a function to get the context, can't the context itself be passed here directly?

JssDWt replied:
Done

guggero commented on routing/control_tower.go:
```diff
@@ -266,6 +279,8 @@ func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
 // immediately.
 func (p *controlTower) SubscribeAllPayments() (*ControlTowerSubscriber, error) {
+	p.newSubscriptionMtx.Lock()
+	defer p.newSubscriptionMtx.Unlock()
```
nit: add empty line after the locks to visually separate them.

JssDWt replied:
Removed locks

guggero commented on routing/control_tower.go:
```diff
@@ -131,6 +143,8 @@ func (p *controlTower) RegisterAttempt(paymentHash lntypes.Hash,
 
 	p.paymentsMtx.Lock(paymentHash)
+	p.newSubscriptionMtx.RLock()
+	defer p.newSubscriptionMtx.RUnlock()
```
Agree! Defers are processed "Last In, First Out", I assume that's taken into account here?

guggero commented on routing/control_tower.go:
```diff
@@ -110,6 +110,10 @@ type controlTower struct {
+
+	// newSubscriptionMutex makes sure no payment events are missed when
+	// a new subscription is made.
+	newSubscriptionMtx sync.RWMutex
```
I think this commit should come before the release notes and changes to the RPC.

JssDWt replied:
Alright, I will fix that. I added that commit last, because Joost and I were discussing whether or not the locks were necessary. There's a different solution without additional locking, but it needs a comment, as shown here:
https://github.com/lightningnetwork/lnd/pull/6335#discussion_r847158161

If you could take a decision which strategy to follow, I can fix up the commits nicely.

JssDWt replied:
So yeah I removed the locks in the current version of this PR, I can reintroduce them if the concession with duplicates/out-of-order events is not what you'd want.

guggero reviewed (commented):
Thank you for the PR!
I did a first level pass to gauge the size of it. Looks pretty slim, though I want to take a closer look at the possibilities of a deadlock.

We're currently busy getting 0.15.0 out and this PR unfortunately didn't make the cut. So we'll come back to it once the window for 0.16 is open. Thank you for your patience in advance!

JssDWt commented:
I went ahead and removed the hard-to-read locking in favor of a comment in the rpc mether that there could be duplicates and/or out-of-order events at the start of the stream if payments are in-flight while subscribing to this method. Either the client doesn't care about out-of-order and duplicate events or they could make sure the subscription returns before sending out any payments.

Also rebased to master and fixed Oliver's comments.

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@jssdwt, remember to re-request review from reviewers when ready

guggero reviewed (commented):
I finally found some time to review the PR. Sorry for the delay!
The diff looks much simpler than before. I think it makes sense to go with this compromise of potential duplicate/out-of-order events. IMO the use case for this RPC is probably that you start the subscriber as soon as you start your app and then keep it open for the whole duration.

But maybe it would be worth adding a comment somewhere how you would make sure you never miss a payment? E.g. subscribe, wait for all initial state updates to arrive, then list all payments from the database to find potential ones that were started in the time between starting the subscription and receiving all updates?

Ah, and needs a rebase too.

yyforyongyu commented on routing/control_tower.go:
```diff
@@ -294,4 +331,21 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
 		case <-subscriber.quit:
```
Since we are here, does anyone know why we don't remove the subscriber here?

JssDWt replied:
I don't know why exactly. I know the subscriber will be removed when the payment reaches a terminal state. And that the locks on subscribersMtx prevent complete parallel updates of payments. It may get pretty busy around that subscribersMtx, so I could imagine acquiring another lock there is unwanted. but that's a guess.

JssDWt commented on routing/control_tower_test.go:
```diff
@@ -182,6 +183,123 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
 	})
 }
 
+func TestPaymentControlSubscribeAllSuccess(t *testing.T) {
```
Done

JssDWt commented on routing/control_tower_test.go:
```diff
@@ -182,6 +183,123 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
+		info1.PaymentIdentifier, attempt1.AttemptID, &settleInfo1,
+	)
+	require.NoError(t, err)
+	require.Equal(t, settleInfo1, *htlcAttempt1.Settle,
```
Done (I think? The docs are not clear about what to do when the arguments don't fit on a single line. Do we fit as many args as possible per line?)

JssDWt commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -731,19 +735,57 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
 	}
+
+	// Stream updates to the client.
+	err = s.trackPaymentStream(stream.Context(), subscription,
```
Done

JssDWt commented on routing/control_tower.go:
```diff
@@ -255,6 +261,36 @@ func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
 	return subscriber, nil
 }
 
+// SubscribeAllPayments subscribes to updates for all payments. A first update
```
Done

JssDWt commented on docs/release-notes/release-notes-0.15.0.md:
```diff
@@ -302,6 +302,10 @@ from occurring that would result in an erroneous force close.](https://github.co
   output](https://github.com/lightningnetwork/lnd/pull/6528).
 
+* Add [TrackPayments](https://github.com/lightningnetwork/lnd/pull/6335)
+  method to the RPC to allow subscribing to updates from any payment. Similar
```
Done (in the wrong release notes still btw)

JssDWt replied:
Done (in the wrong release notes still btw)

JssDWt commented on lnrpc/routerrpc/router.proto:
```diff
@@ -22,6 +22,14 @@ service Router {
+    subscription, the start of the payment stream could produce out-of-order
+    and/or duplicate events.
+    */
+    rpc TrackPayments (TrackPaymentsRequest) returns (stream lnrpc.Payment);
```
Done

JssDWt commented on routing/control_tower_test.go:
```diff
@@ -182,6 +183,123 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
+	err = pControl.RegisterAttempt(info1.PaymentIdentifier, attempt1)
+	require.NoError(t, err)
+
+	// Initiate a second payment after the subscription is already active
```
Done

JssDWt commented on routing/control_tower.go:
```diff
@@ -57,6 +57,11 @@ type ControlTower interface {
+	// SubscribeAllPayments subscribes to updates for all payments. A first
+	// update with the current state of every inflight payment is always
+	// sent out immediately.
+	SubscribeAllPayments() (*ControlTowerSubscriber, error)
```
Done

positiveblue commented on routing/control_tower.go:
```diff
@@ -305,4 +344,21 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
+		// list.
+		case <-subscriber.quit:
+			p.subscribersMtx.Lock()
+			p.subscribersAllPayments = append(
```
I am a bit worried about this. We are modifying the array size here but `range` is evaluated only once (at the beginning of the loop, so if you change it's size it will not be reflected in the next iteration).

An [example with playground](https://go.dev/play/p/YpsrnMS8ARX) to see the real behavior + the definition from the [go spec](https://go.dev/ref/spec#For_range) in case I am missing something.

JssDWt replied:
Fixed this by using a map and removing based on the key. Should be safe according to [stackoverflow](https://stackoverflow.com/a/23230406/3883866)

yyforyongyu commented on routing/control_tower.go:
```diff
@@ -102,8 +107,9 @@ func (s *ControlTowerSubscriber) Close() {
 
-	subscribers    map[lntypes.Hash][]*ControlTowerSubscriber
-	subscribersMtx sync.Mutex
+	subscribersAllPayments []*ControlTowerSubscriber
```
So I went to check the unique ID mentioned in the review club. Turned out other implementations involve a db write, such as `htlcswitch/sequencer`, which could be an overkill here. So I think we could either implement a monotonically increasing integer here as our unique ID, or use `uuid`(which seems to be much simpler) if we decide to replace the slice with a map here.

JssDWt replied:
Thanks for checking! I noticed uuid is not used anywhere else in the codebase. So maybe the monotonically increasing integer would be the best way to go. If you'd not reboot LND for one year, you could do 584942417 unique updates per millisecond. Which is a ridiculous number, so a uint64 should be safe to use here.

yyforyongyu replied:
>  If you'd not reboot LND for one year, you could do 584942417 unique updates per millisecond.

Interesting. How did you come up with this figure?

> Which is a ridiculous number, so a uint64 should be safe to use here.

I don't think the number of the updates matter here since the ID is used to track the number of payments subscriptions? Ideally a single all-payments subscriber should be enough to fit all.

JssDWt replied:
> Interesting. How did you come up with this figure?
I was just checking how many unique numbers you can create with this and how many subscriptions you could do in the time LND is running. I was thinking if you have some weird subscription client that keeps subscribing/unsubscribing, how long will it take until the number rolls over. The answer is: ridiculously long, basically. That's where that came from. It's ulong64 maxvalue / number of milliseconds in a year.

> I don't think the number of the updates matter here
So yeah that was about how many updates to the subscriptions were made.

JssDWt commented on routing/control_tower_test.go:
```diff
@@ -182,6 +183,123 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
+	err = pControl.InitPayment(info1.PaymentIdentifier, info1)
+	require.NoError(t, err)
+
+	// Subscription should succeed and immediately report the InFlight
```
Added a test for this

JssDWt commented on routing/control_tower_test.go:
```diff
@@ -182,6 +183,123 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
+		results[info1.PaymentIdentifier].Status == sif ||
+		results[info2.PaymentIdentifier].Status == sif {
+		select {
+		case item := <-subscription.Updates:
```
Reading it exactly 5 times now

JssDWt commented:
I've added a bunch of things
- itest
- test for the routerrpc method
- test for immediate notification of inflight payments when starting a subscription when payments are inflight
- map instead of slice to keep the subscriptions, with a unique id in order to easily remove them (and a test for that)

yyforyongyu commented on routing/control_tower_test.go:
```diff
@@ -179,6 +179,244 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
+}
+
+// TestPaymentControlSubscribeAllImmediate tests whether already inflight
+// payments are reported at the start of the SubscribeAllPayments subscription.
```
👍

yyforyongyu commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -737,19 +741,56 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+
+// TrackPayments returns a stream of payment state updates.
+func (s *Server) TrackPayments(request *TrackPaymentsRequest,
+	stream Router_TrackPaymentsServer) error {
```
Can we use `Router_TrackPaymentsV2Server` here? `Router_TrackPaymentsServer` is deprecated.

JssDWt replied:
Do you mean rename the newly created method `TrackPayments` to `TrackPaymentsV2`? I think you may be confused with `TrackPayment` (without the `s`)?

yyforyongyu replied:
hah yeah my bad! Didn't see the `s`.

yyforyongyu commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -731,19 +735,57 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
+}
+
+// trackPaymentStream streams payment updates to the client.
+func (s *Server) trackPaymentStream(context context.Context,
```
I think `Router_TrackPaymentsServer` is deprecated. There's a `TrackPayment` inside `router_server_deprecated.go` where we actually convert it to the newer version using `legacyTrackPaymentServer`.

yyforyongyu commented on lnrpc/routerrpc/router_server.go:
```diff
@@ -760,23 +802,24 @@ func (s *Server) trackPayment(identifier lntypes.Hash,
-			log.Debugf("Payment status stream %v canceled", identifier)
-			return stream.Context().Err()
+		case <-context.Done():
+			return context.Err()
```
nit: debug log is not added back🧐

yyforyongyu commented on lntest/itest/lnd_trackpayments_test.go:
```diff
@@ -0,0 +1,108 @@
+// receives payment updates.
+func testTrackPayments(net *lntest.NetworkHarness, t *harnessTest) {
+	// Setup two connected nodes.
+	alice := net.NewNode(t.t, "alice", nil)
```
nit: could use `net.Alice` and `net.Bob` instead, which saves us from sending coins, making connections, etc.

JssDWt replied:
Done. I hope the cleanup is good, it's only closing the channel between `net.Alice` and `net.Bob`

yyforyongyu replied:
Yeah it actually just creates Alice and Bob every time a new case runs, and I'm fixing the itest.

yyforyongyu commented on lntest/itest/lnd_trackpayments_test.go:
```diff
@@ -0,0 +1,108 @@
+
+// testTrackPayments tests whether a client that calls the TrackPayments api
+// receives payment updates.
+func testTrackPayments(net *lntest.NetworkHarness, t *harnessTest) {
```
👍

yyforyongyu commented on docs/release-notes/release-notes-0.15.1.md:
```diff
@@ -69,6 +69,10 @@
   flow](https://github.com/lightningnetwork/lnd/pull/6754) so that the server 
   can indicate to the client that it has completed the RPC MW registration.
 
+* Add [TrackPayments](https://github.com/lightningnetwork/lnd/pull/6335)
```
As pointed out by @guggero, we could aim this for `v0.16.0`.

JssDWt replied:
Moved to release notes of 0.16.0

yyforyongyu commented on lnrpc/routerrpc/router_server_test.go:
```diff
@@ -0,0 +1,169 @@
+package routerrpc
```
Dope! We usually test RPC changes in itest, while itself is missing unit tests, so this is great!

JssDWt replied:
Yeah it's a bit easier to get the backend to give you what you want to test here. Otherwise you're dependent on pathfinding for example to test this method, which makes it a bit complicated.

yyforyongyu replied:
Exactly! I really like that we have this controlled environment to check the expected state changes in our system.

yyforyongyu commented on lnrpc/routerrpc/router_server_test.go:
```diff
@@ -0,0 +1,169 @@
+	}
+}
+
+type StreamMock struct {
```
We now use `testify/mock` in our tests, check `routing/mock_test.go`. I think we could export the mockers there.

JssDWt replied:
There's one mock for the control tower and one for the rpc stream. Should I move the one for the control tower to `routing/mock_test.go` and create a new `lnrpc/routerrpc/mock_test.go`?

yyforyongyu replied:
We already have `mockControlTower` in `routing/mock_test.go`. So I think you could export that, renaming it to `MockControlTower` and use it here. As for the rpc stream, yeah, leave it here.

JssDWt replied:
That `mockControlTower` is in the `routing` package. The test for the `router_server` I made is in the `routerrpc` package. I can't import a type from another package in the tests (see [here](https://stackoverflow.com/a/22277807/3883866)). 

So there's 3 options, I think.
1) Leave the code as it is now, with a new mock for this specific test.
2) Always compile `mock_test` into the `routing` package.
3) Create a seperate package for the mocks in `mock_test`

I obviously prefer option 1, because that's no work, lol.
What do you think?

yyforyongyu replied:
Oh yeah forgot about that. In that case I'd prefer option 1 for the moment, just to keep the PR concise. Once we start patching unit tests for RPCs, the mockers could live inside `lntest/mock`, which I believe was intended for such a job.

yyforyongyu reviewed (changes requested):
Great work @JssDwt! Thanks for making the effort to write the tests which are very thorough. I think this is getting close!

JssDWt commented on routing/control_tower.go:
```diff
@@ -289,8 +335,8 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
-	// Notify all subscribers of the event.
-	for _, subscriber := range list {
+	// Notify all subscribers that subscribed to the current payment hash.
+	for _, subscriber := range subscribersPaymentHash {
```
Yes exactly

JssDWt commented:
@yyforyongyu You mentioned this:

> nit: debug log is not added back🧐

I did add it back, but it's one level higher in the callstack:
- [here](https://github.com/lightningnetwork/lnd/pull/6335/files#diff-9ced6590b7c4cfad60e1fa1641b602da6fe758092211f30a55b19771e4186842R759)
- [and here](https://github.com/lightningnetwork/lnd/pull/6335/files#diff-9ced6590b7c4cfad60e1fa1641b602da6fe758092211f30a55b19771e4186842R786)

JssDWt commented on routing/control_tower.go:
```diff
@@ -114,7 +124,9 @@ type controlTower struct {
-		db:          db,
+		db: db,
+		subscribersAllPayments: make(
+			map[uint64]*ControlTowerSubscriber),
```
Oops! Fixed.
I might work on something to make the linter catch these formatting errors.

yyforyongyu replied:
Dope! Something we all want!

yyforyongyu reviewed (approved):
LGTM🎉 Thanks again for the great work @JssDWt !

lightninglabs-deploy commented:
@guggero: review reminder
@positiveblue: review reminder
@jssdwt, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero: review reminder
@positiveblue: review reminder
@jssdwt, remember to re-request review from reviewers when ready

guggero commented on routing/control_tower.go:
```diff
@@ -266,6 +279,39 @@ func (p *controlTower) SubscribePayment(paymentHash lntypes.Hash) (
+
+	for _, payment := range inflightPayments {
+		// Always write current payment state to the channel.
+		subscriber.queue.ChanIn() <- payment
```
Using loop variables that are pointers is tricky in Golang. It can happen that you end up with the same (last) element being used everywhere if the variable isn't put on the stack (which I'm not sure is the case when sending to a channel). Probably best to use an index based loop here and then just send `inflightPayments[idx]` to avoid the issue.

JssDWt replied:
Done

guggero commented on routing/control_tower.go:
```diff
@@ -305,4 +352,18 @@ func (p *controlTower) notifySubscribers(paymentHash lntypes.Hash,
 	}
+
+	// Notify all subscribers that subscribed to all payments.
+	for key, subscriber := range p.subscribersAllPayments {
```
We still read from a map that could be written to concurrently. I think we need to hold the lock, create a local copy, then release the lock again and loop over the copy instead. Then only lock again on delete (as it's currently implemented on L364 and following).

positiveblue replied:
We can see the problem [here](https://go.dev/play/p/eDysaEIMGGA)

JssDWt replied:
I do agree this is the case. But there's a note on the rpc call saying:

> Note that if payments are in-flight while starting a new subscription, the start of the payment stream could produce out-of-order and/or duplicate events. In order to get updates for every in-flight payment attempt make sure to subscribe to this method before initiating any payments.

That effect doesn't change whether we hold the lock and copy the map locally or not. It will still depend on timing whether you receive a duplicate or not. (Because the inflight HTLCs are fetched too when you subscribe to trackpayments)

So I'd prefer to not copy the thing over to save some cpu and time in a global lock.

positiveblue replied:
This could potentially end up in a panic by the go runtime (Concurrent Map Misuse)

```
fatal error: concurrent map iteration and map write
```

To make it even worst, this is a `throw` not a `panic` which means there is no way to recover. The only solution here is to actually copy the map and iterate or use a lock.

JssDWt replied:
OK! I'm convinced. Done.

guggero commented on routing/control_tower_test.go:
```diff
@@ -179,6 +179,244 @@ func TestPaymentControlSubscribeFail(t *testing.T) {
+	)
+
+	htlc1 := result1.HTLCs[0]
+	require.Truef(
```
nit: could use `require.Equal()` which does a deep equal and also nice diff view (so the `spew.Sdump` isn't needed necessarily).

JssDWt replied:
Done

guggero reviewed (commented):
We're getting really close here!

positiveblue commented on lnrpc/routerrpc/router_server_test.go:
```diff
@@ -0,0 +1,169 @@
+
+	// Listen to payment updates in a goroutine.
+	go func() {
+		err := server.TrackPayments(request, stream)
```
q: what happens to this go routine after the test is done?

JssDWt replied:
Well that was more work than I expected...

Indeed that go routine would last forever I think (until the test run was done probably)
So I decided to pass a cancellable context to that stream so the trackpayments call would complete after the test was complete.

Turns out that caused a panic, because the ControlTowerSubscriber tries to close the `quit` chan when the TrackPayments call completes. So I had to convert ControlTowerSubscriber to an interface in order to be able to mock it in the test.

Did these things in a newly added commit.

positiveblue replied:
Sorry for that :pray:

positiveblue reviewed (changes requested):
Really nice, this feature will be pretty useful. I think this only needs a couple of changes and we are ready to go (pun intended).

The blocking one is the one mentioned by @guggero  we need to hold a lock whenever we are accessing/changing the map and that also applies to for iterations with the `range` keyword,

The rest is :ok_hand:

lightninglabs-deploy commented:
@jssdwt, remember to re-request review from reviewers when ready

JssDWt commented:
I think it's not my fault the Travis CI and itests are failing?
Looks like travis CI needs to be updated to have go 1.19.0
And the itests are a bit flaky maybe?
Or is it my fault?

positiveblue commented:
@JssDWt yes, some tests are flaky + the travis thing is not your fault. There is already a [PR](https://github.com/lightningnetwork/lnd/pull/6893) fixing the travis ci step + the steps that are a must are the ones makreds as `required`

JssDWt commented on routing/control_tower.go:
```diff
@@ -60,40 +60,46 @@ type ControlTower interface {
 	// received.
-	Updates <-chan interface{}
+	Updates() <-chan interface{}
+	Close()
```
Oops, sorry about that. Fixed.

JssDWt commented on routing/control_tower.go:
```diff
@@ -102,6 +108,10 @@ func (s *ControlTowerSubscriber) Close() {
 	s.queue.Stop()
 }
 
+func (s *controlTowerSubscriberImpl) Updates() <-chan interface{} {
```
Oops, sorry about that. Fixed.

positiveblue reviewed (approved):
Ready to be merged :rocket: thank you for this @JssDWt

JssDWt commented:
Thanks for the reviews and your patience guys! You rock.
